rtos/ConditionVariable.h

```diff
@@ -0,0 +1,186 @@
+ * mutex after ConditionVariable::notify_all is called is undefined.
+ * * When ConditionVariable::notify_one or ConditionVariable::notify_all is
+ * called and there are one or more waiters and one or more threads attempting
+ * to acquire the condition variable's mutex the order in which the mutex is
+ * acquired is undefined.
+ * * The behavior of ConditionVariable::wait and ConditionVariable::wait_for
+ * is undefined if the condition variable's mutex is locked more than once by
+ * the calling thread.
+ * * Spurious notifications (not triggered by the application) can occur
+ * and it is not defined when these occur.
+ *
+ * @note Synchronization level: Thread safe
+ *
+ * Example:
+ * @code
+ * #include "mbed.h"
+ *
+ * Mutex mutex;
+ * ConditionVariable cond(mutex);
+ *
+ * // These variables are protected by locking mutex
+ * uint32_t count = 0;
+ * bool done = false;
+ *
+ * void worker_thread()
+ * {
+ *     mutex.lock();
+ *     do {
+ *         printf("Worker: Count %lu\r\n", count);
+ *
+ *         // Wait for a condition to change
+ *         cond.wait();
+ *
+ *     } while (!done);
+ *     printf("Worker: Exiting\r\n");
+ *     mutex.unlock();
+ * }
+ *
+ * int main() {
+ *     Thread thread;
+ *     thread.start(worker_thread);
+ *
+ *     for (int i = 0; i < 5; i++) {
+ *
+ *         mutex.lock();
+ *         // Change count and notify waiters of this
+ *         count++;
+ *         printf("Main: Set count to %lu\r\n", count);
+ *         cond.notify_all();
+ *         mutex.unlock();
+ *
+ *         wait(1.0);
+ *     }
+ *
+ *     mutex.lock();
+ *     // Change done and notify waiters of this
+ *     done = true;
+ *     printf("Main: Set done\r\n");
+ *     cond.notify_all();
+ *     mutex.unlock();
+ *
+ *     thread.join();
+ * }
+ * @endcode
+ */
+class ConditionVariable : private mbed::NonCopyable<ConditionVariable> {
+public:
+    /** Create and Initialize a ConditionVariable object */
+    ConditionVariable(Mutex &mutex);
+
+    /** Wait for a notification
+     *
+     * Wait until a notification occurs.
+     *
+     * @note - The thread calling this function must be the owner of the
+     * ConditionVariable's mutex and it must be locked once and only once
+     *
+     * Example:
+     * @code
+     * mutex.lock();
+     * while (!condition_met) {
+     *     timeout = cond.wait();
+     * }
+     * mutex.unlock();
+     * @endcode
+     */
+    void wait();
+
+    /** Wait for a notification or timeout
+     *
+     * @param   millisec  timeout value or 0 in case of no time-out.
+     * @return  true if woken from notification, false if timed out.
+     *
+     * @note - The thread calling this function must be the owner of the
+     * ConditionVariable's mutex and it must be locked once and only once
+     * @note - Spurious notifications can occur so the caller of this API should check the
+     *
+     * Example:
+     * @code
+     * mutex.lock();
+     * bool woke_from_notify = false;
```

Comment:
cf comments above - "woke from notify" is misleading - it's really "timeout not expired". Also, this doesn't initialise it correctly for the loop to work.

Reply:
done